From b3d31f5772f2c103e602327834824817fee54845 Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Fri, 10 Oct 2014 21:03:32 -0400
Subject: [block] blk-mq: remove unnecessary blk_clear_rq_complete()

Message-id: <1412975015-5370-32-git-send-email-jmoyer@redhat.com>
Patchwork-id: 97459
O-Subject: [RHEL7 PATCH 31/34] blk-mq: remove unnecessary blk_clear_rq_complete()
Bugzilla: 1146660
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Mike Snitzer <snitzer@redhat.com>

This is a backport of the following commit.  This resolves bug
1146660.

  commit 2edd2c740b2918eb0a9a1fe1b69678b903769ec2
  Author: Ming Lei <ming.lei@canoical.com>
  Date:   Wed Sep 17 17:47:58 2014 +0800

    blk-mq: remove unnecessary blk_clear_rq_complete()

    This patch removes two unnecessary blk_clear_rq_complete(),
    the REQ_ATOM_COMPLETE flag is cleared inside blk_mq_start_request(),
    so:

     - The blk_clear_rq_complete() in blk_flush_restore_request()
     needn't because the request will be freed later, and clearing
     it here may open a small race window with timeout.

     - The blk_clear_rq_complete() in blk_mq_requeue_request() isn't
     necessary too, even though REQ_ATOM_STARTED is cleared in
     __blk_mq_requeue_request(), in theory it still may cause a small
     race window with timeout since the two clear_bit() may be
     reordered.

    Signed-off-by: Ming Lei <ming.lei@canoical.com>
    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-flush.c b/block/blk-flush.c
index 68f9643..3c1c7f4 100644
--- a/block/blk-flush.c
+++ b/block/blk-flush.c
@@ -126,8 +126,6 @@ static void blk_flush_restore_request(struct request *rq)
  /* make @rq a normal request */
  rq->cmd_flags &= ~REQ_FLUSH_SEQ;
  rq->end_io = rq->flush.saved_end_io;
-
- blk_clear_rq_complete(rq);
 }
 
 static bool blk_flush_queue_rq(struct request *rq, bool add_front)
diff --git a/block/blk-mq.c b/block/blk-mq.c
index d9c6634..be24273 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -438,7 +438,6 @@ static void __blk_mq_requeue_request(struct request *rq)
 void blk_mq_requeue_request(struct request *rq)
 {
  __blk_mq_requeue_request(rq);
- blk_clear_rq_complete(rq);
 
  BUG_ON(blk_queued_rq(rq));
  blk_mq_add_to_requeue_list(rq, true);
-- 
1.7.1